### PR TITLE
add external memory network demo

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This PR includes an example implementation of an external memory network and example usage with a simple task.

### CODE REVIEW

File: demo/memnet/external_memory.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        memory_remove = mixed_layer(input=dotmul_operator(a=self.external_memory, b=out_prod))
+
+        memory_remove_neg = slope_intercept_layer(input=memory_remove, slope=-1.0, intercept=0)
+
+        # memory_updated = memory_mat - memory_remove  = memory_mat + memory_remove_neg
+        memory_removed = mixed_layer(input = [identity_projection(input=self.external_memory),
+                                              identity_projection(input=memory_remove_neg)],
+                                     bias_attr = False,
+                                     act = LinearActivation())
```
Comment (emailweixu) [inline demo/memnet/external_memory.py]: 78 and 81 can be combinded as written as: memory_removed = self.external_memory - memory_remove.
See https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/tests/configs/math_ops.py

File: demo/memnet/external_memory.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                              name='add_vector')
+
+        erase_vec = self.MakeConstantVector(self.mem_fea_size, 1.0, write_key)
+
+
+        if self.is_test:
+            print_layer(input=[erase_vec])
+            print_layer(input=[add_vec])
+
+        out_prod = out_prod_layer(norm_cosine_similarity_write, erase_vec, name="outer")
```
Comment (emailweixu) [inline demo/memnet/external_memory.py]: Creating a constant vector erase_vec for this is very ugly. A nicer way to do this is to enhance "repeat" layer to allow repeat in both directions, similar to "repmat" in matlat.

File: demo/memnet/external_memory.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        # memory_updated = memory_mat - memory_remove  = memory_mat + memory_remove_neg
+        memory_removed = mixed_layer(input = [identity_projection(input=self.external_memory),
+                                              identity_projection(input=memory_remove_neg)],
+                                     bias_attr = False,
+                                     act = LinearActivation())
+
+        out_prod_add = out_prod_layer(norm_cosine_similarity_write, add_vec, name="outer_add")
+
+        memory_output = mixed_layer(input = [identity_projection(input=memory_removed),
```
Comment (emailweixu) [inline demo/memnet/external_memory.py]: Using addto_layer can make this looks simpler.

File: demo/memnet/external_memory.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from paddle.trainer_config_helpers import *
+
+
+class ExternalMemory(object):
```
Comment (emailweixu) [inline demo/memnet/external_memory.py]: need to comment for the class, and its member functions.

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: demo/memnet/external_memory.py (hunk 4 above)
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: Comments have been added to both the class and member functions.

File: demo/memnet/external_memory.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    def __init__(self, name, mem_slot_size, mem_fea_size, ext_mem_initial, is_test=False, scale=5):
+        self.name = name
+        self.mem_slot_size = mem_slot_size
+        self.mem_fea_size = mem_fea_size
+        self.scale = 5
```
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: This has been corrected.

File: demo/memnet/external_memory.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        self.external_memory = memory(name=self.name,
+                               	size=mem_fea_size*mem_slot_size,
+                                boot_bias= ParamAttr(initial_std=0.01,
```
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: This has been updated.

File: demo/memnet/external_memory.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        self.is_test = is_test
+
+    def read(self, read_key):
+        cosine_similarity_read = cos_sim(read_key, self.external_memory, scale=self.scale, size=self.mem_slot_size)
+        norm_cosine_similarity_read = mixed_layer(input=
+                                                  identity_projection(cosine_similarity_read),
+                                                  bias_attr = False,
+                                                  act = SoftmaxActivation(),
+                                                  size = self.mem_slot_size,
+                                                  name='read_weight')
```
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: Similar issues have been addressed.

File: demo/memnet/external_memory.py (hunk 1 above)
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: This part of the code has been updated using math_ops

File: demo/memnet/external_memory.py (hunk 3 above)
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: Switch to addto_layer

File: demo/memnet/external_memory.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                                             identity_projection(input=out_prod_add)],
+                                    bias_attr = False,
+                                    act = LinearActivation(),
+                                    name=self.name)
+        if self.is_test:
+            print_layer(input=[memory_output])
+
+        return memory_output
+
+    def MakeConstantVector(self, vec_size, value, dummy_input):
```
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: Changed the function name following the convention.

File: demo/memnet/external_memory.py (hunk 2 above)
Comment (Haichao-Zhang) [inline demo/memnet/external_memory.py]: Looking into the repeat layer currently.

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献代码。由于Paddle V1/V2版本已不再维护，相关代码也已从develop分支上删除，因此关闭您的PR，欢迎您向Paddle最新版-Fluid贡献代码。
Thanks for contributing to PaddlePaddle! Since V1/V2 will not be maintained anymore, and related codes have been deleted from develop branch as well, we close this PR. Welcome to contribute to Fluid——the latest version of PaddlePaddle.

**Pull request closed without merging**